src/Script/Script.ts

```diff
@@ -75,44 +52,40 @@ export class Script {
     }, 0);
   }
 
-  /**
-   * Marks this script as having been updated. It will be recompiled next time something tries
-   * to exec it.
-   */
-  markUpdated(): void {
-    this.module = null;
-    this.moduleSequenceNumber = ++globalModuleSequenceNumber;
+  /** Invalidates the current script module and related data, e.g. when modifying the file. */
+  invalidateModule(): void {
+    delete this.module;
+    if (this.url) URL.revokeObjectURL(this.url);
+    delete this.url;
+    delete this.ramUsage;
+    delete this.ramUsageEntries;
+    for (const [url, dependency] of this.dependencies) {
+      dependency.dependents.delete(this);
```

Comment:
Something subtle that is probably worth documenting: This is over-eager. Because the dependencies are both direct *and* indirect, this loop can remove items from other script's `dependents` when those items are still dependent through other paths.

I *think*, but am not 100% certain, that this is OK because it will be fixed up when `compile` is run.

Reply:
The only thing being removed from other scripts' dependents are scripts that are currently being invalidated, which will only be the edited script and anything higher in the import tree. No scripts that remain valid will be removed as dependents.

Dependents is only used for passing invalidation up the import tree, so there is no need or desire to store dependents for any scripts that are already invalidated.

Reply:
I did change up the invalidation function a little in f86f6f5 to prevent re-invalidating the same script a bunch of times in instances where there's a complex import tree. Now there's an early return if the script has already been invalidated, and also over in NetscriptJSEvaluator we're now only storing direct dependents instead of deep-dependents. Since invalidation is passed to all dependents, it will naturally reach all deep dependents anyway, this just limits the amount that we try invalidating the same script multiple times when a file is edited.

Reply:
Resolving this one, but if you do find that there's a failure case I'm just not seeing, either reopen or just start a new thread